diatheke/audio_client.cpp

```diff
@@ -0,0 +1,206 @@
+
+  // Record until we get a result
+  DiathekePB::ASRResult result =
+      Diatheke::ReadASRAudio(stream, &recorder, 8192);
+  recorder.stop();
+
+  // Display the result
+  std::cout << "\n  ASRResult:" << std::endl;
+  std::cout << "    Text: " << result.text() << std::endl;
+  std::cout << "    Confidence: " << result.confidence() << std::endl;
+
+  return client->processASRResult(session.token(), result);
+}
+
+// Uses TTS to play back the reply as speech.
+void handleReply(Diatheke::Client *client,
+                 const DiathekePB::ReplyAction &reply) {
+  std::cout << "\n  Reply:" << std::endl;
+  std::cout << "    Text: " << reply.text() << std::endl;
+  std::cout << "    Luna Model: " << reply.luna_model() << std::endl;
+
+  // Create the TTS stream
+  Diatheke::TTSStream stream = client->newTTSStream(reply);
+
+  // Start the player
+  Player player(playCmd);
+  player.start();
+  Diatheke::WriteTTSAudio(stream, &player);
+  player.stop();
+}
+
+/*
+ * Executes the task specified by the given command and
+ * returns an updated session based on the command result.
+ */
+DiathekeSession handleCommand(Diatheke::Client *client,
+                              const DiathekeSession &session,
+                              const DiathekePB::CommandAction &cmd) {
+  // Print the command info
+  std::cout << std::endl;
+  std::cout << "  Command:" << std::endl;
+  std::cout << "    ID:" << cmd.id() << std::endl;
+  std::cout << "    Input params:" << std::endl;
+  auto params = cmd.input_parameters();
+  for (auto iter = params.begin(); iter != params.end(); iter++) {
+    std::cout << "      " << iter->first << " = " << iter->second << std::endl;
+  }
+
+  // Update the session with the command result
+  DiathekePB::CommandResult result;
+  result.set_id(cmd.id());
+  return client->processCommandResult(session.token(), result);
+}
+
+/*
+ * Executes the actions for the given session and returns
+ * an updated session.
+ */
+DiathekeSession processActions(Diatheke::Client *client,
+                               const DiathekeSession &session) {
+  // Iterate through each action in the list and determine its type.
+  for (auto action : session.action_list()) {
+    if (action.has_input()) {
+      // The WaitForUserAction will involve a session update.
+      return waitForInput(client, session, action.input());
+    } else if (action.has_reply()) {
+      // Replies do not require a session update.
+      handleReply(client, action.reply());
+    } else if (action.has_command()) {
+      // The CommandAction will involve a session update.
+      return handleCommand(client, session, action.command());
+    } else {
+      throw std::runtime_error("received unknown action type");
+    }
+  }
+
+  throw std::runtime_error(
+      "ran all actions without receiving an updated session");
+}
+
+int main(int argc, char *argv[]) {
+  // Create the client
+  Diatheke::Client client(serverAddress, insecureConnection);
+
```

Comment:
Might want to add a try - catch here. I was trying to connect to a secure server with the insecure flag on and it raised a Diatheke::ClientError
```suggestion
  try {
    // Create the client
    Diatheke::Client client(serverAddress, insecureConnection);
  } catch (const std::exception &e) {
    std::cout << "Error connecting to server: " << e.what() << std::endl;
    return 1;
  }

```

Reply:
I just moved everything into the try/catch block. It needed to be done for the ListModels and Version methods, too.